docs/architecture/0e-plans.md

```diff
@@ -0,0 +1,474 @@
+
+```
+aes.h
+aria.h
+bignum.h
+block_cipher.h
+camellia.h
+ccm.h
+chacha20.h
+chachapoly.h
+cipher.h
+cmac.h
+des.h
+ecdh.h
+ecdsa.h
+ecjpake.h
+ecp.h
+gcm.h
+md.h
+md5.h
+poly1305.h
+ripemd160.h
+sha1.h
+sha256.h
+sha3.h
+sha512.h
+```
+
+Main loss of functionality:
+
+* Self-test functions. See TODO
+* Access to bignum and ECC arithmetic. We've decided that this is acceptable.
+
+#### Headers that can be made fully private
+
+The headers listed below are not used in Mbed TLS, except in places that should be removed and can be removed easily.
+
+```
+dhm.h
+hmac_drbg.h
+pkcs12.h
+pkcs5.h
+rsa.h
+```
+
+Places where some of these headers are used:
+
+* `programs/fuzz/fuzz_*key.c` (to fuzz RSA functions that are now private)
+* `programs/ssl/ssl_test_lib.c` (HMAC\_DRBG; can move to only using the PSA RNG)
+* `programs/test/benchmark.c` (HMAC\_DRBG)
+* `programs/test/selftest.c`
+* `scripts/data_files/query_config.fmt`
+
+Main loss of functionality:
+
+* Finite-field Diffie-Hellman with arbitrary groups. We've decided that this is acceptable.
+* Custom RSA mechanisms. We've decided that this is acceptable.
+* PKCS5 and PKCS12 mechanisms except as exposed by the pk module. We've decided that this is acceptable.
+* HMAC\_DRBG in itself (i.e. outside of deterministic ECDSA and for the PSA Crypto RNG instance). We intend to restore this functionality through a PSA API, but the API isn't designed yet, so this will happen after 1.0 and not with the existing API.
+
+#### Headers that will become private eventually
+
+The headers listed below should be private, but are currently used in Mbed TLS to an extent that makes it hard to remove before the 1.0/4.0 release. As a result, they need to remain visible to Mbed TLS, but should be clearly indicated as not part of the stable API.
+
+```
+ctr_drbg.h
+entropy.h
+error_common.h
+oid.h
+```
+
+Main loss of functionality:
+
+* CTR\_DRBG in itself (i.e. other than for the PSA Crypto RNG instance). We intend to restore this functionality through a PSA API, but the API isn't designed yet, so this will happen after 1.0 and not with the existing API.
+* Direct access to entropy sources. We've decided that this is acceptable.
+* The ability to configure entropy sources on a platform. This is not an acceptable loss. In the long term (likely after 1.0), this will be resolved by the PSA crypto random driver API. In the short term, we will expose a modified `mbedtls_hardware_poll()` (https://github.com/Mbed-TLS/mbedtls/issues/9618).
+
+#### Privatization of Everest headers
+
+Everest headers (`drivers/everest/include/everest/include/**/*.h`) contain some exposed types: they are exposed via `mbedtls_ecdh_context` from `mbedtls/ecdh.h` which is exposed via `mbedtls_psa_key_agreement_interruptible_operation_t` indirectly from `psa/crypto.h`. The rest of their content is private (to be consumed only by `ecdh.c`) or internal (to be consumed only by `everest/**/*.c`) definitions.
```

Comment:
Off-topic: in the long run, the built-in driver for ECDH (interruptible or not) should really call `ecp_mul` directly, just like the built-in driver for FFDH calls `exp_mod` directly, and then we just remove `ecdh.c` and `ecdh.h`.